sql/analyzer/resolve_column_defaults.go

```diff
@@ -586,37 +607,133 @@ func resolveColumnDefaults(ctx *sql.Context, a *Analyzer, n sql.Node, scope *Sco
+	return node, identity, err
+}
+
+// transformColumnDefaultsForNode walks over the expressions contained in the specified node and transforms all
+// UnresolvedColumnDefaultValues into ColumnDefaultValues.
+func transformColumnDefaultsForNode(ctx *sql.Context, input sql.Node) (sql.Node, transform.TreeIdentity, error) {
+	return transform.NodeExprsWithNode(input, func(node sql.Node, e sql.Expression) (sql.Expression, transform.TreeIdentity, error) {
 		eWrapper, ok := e.(*expression.Wrapper)
 		if !ok {
 			return e, transform.SameTree, nil
 		}
-		switch n.(type) {
-		case *plan.InsertDestination, *plan.AddColumn, *plan.ShowColumns, *plan.ShowCreateTable, *plan.RenameColumn, *plan.ModifyColumn, *plan.DropColumn, *plan.CreateTable:
-			n, same, err := parseColumnDefaultsForWrapper(ctx, eWrapper)
-			return n, same, err
+		switch node.(type) {
+		case *plan.Values, *plan.InsertDestination, *plan.AddColumn, *plan.ShowColumns, *plan.ShowCreateTable, *plan.RenameColumn, *plan.ModifyColumn, *plan.DropColumn, *plan.CreateTable:
+			return parseColumnDefaultsForWrapper(ctx, eWrapper)
 		default:
 			return e, transform.SameTree, nil
 		}
 	})
 }
 
+// fillInColumnDefaults fills in column default expressions for any source data that explicitly used
+// the 'DEFAULT' keyword as input. This requires that the InsertInto Destination has been fully resolved
+// before the column default values can be used.
+func fillInColumnDefaults(_ *sql.Context, insertInto *plan.InsertInto) error {
+	schema := insertInto.Destination.Schema()
+
+	// If no column names were specified in the query, go ahead and fill
+	// them all in now that the destination is resolved.
+	if len(insertInto.ColumnNames) == 0 {
+		insertInto.ColumnNames = make([]string, len(schema))
+		for i, col := range schema {
+			insertInto.ColumnNames[i] = col.Name
+		}
+	}
+
+	// If the source values are not specified, fill in column default placeholders
+	if values, ok := insertInto.Source.(*plan.Values); ok {
+		for i, tuple := range values.ExpressionTuples {
+			if len(tuple) == 0 {
+				tuple = make([]sql.Expression, len(insertInto.ColumnNames))
+				for j, _ := range tuple {
+					tuple[j] = expression.NewDefaultColumn(insertInto.ColumnNames[j])
+				}
+				values.ExpressionTuples[i] = tuple
+			}
+		}
+	}
+
+	// Pull the column default values out into the same order the columns were specified
+	columnDefaultValues := make([]*sql.ColumnDefaultValue, len(insertInto.ColumnNames))
+	for i, columnName := range insertInto.ColumnNames {
+		found := false
+		for _, col := range schema {
+			if strings.ToLower(col.Name) == strings.ToLower(columnName) {
+				columnDefaultValues[i] = col.Default
+				found = true
+				break
+			}
+		}
+		if !found {
+			return plan.ErrInsertIntoNonexistentColumn.New(columnName)
+		}
+	}
+
+	if insertInto.Source == nil {
+		return fmt.Errorf("no source specified for insert into statement")
```

Comment:
Is this possible?

Reply:
I think you're right... all the uses of `insertInto.Source` are guarded by type checks, so this shouldn't be necessary.